Sources/MongoSwift/MongoCollection+Write.swift

```diff
@@ -428,8 +362,38 @@ public struct UpdateResult: Decodable {
     public let modifiedCount: Int
 
     /// The identifier of the inserted document if an upsert took place.
-    public let upsertedId: AnyBSONValue?
+    public let upsertedId: BSONValue?
 
     /// The number of documents that were upserted.
     public let upsertedCount: Int
+
+    internal init?(from result: BulkWriteResult?) throws {
+        guard let result = result else {
+            return nil
+        }
+        self.matchedCount = result.matchedCount
+        self.modifiedCount = result.modifiedCount
+        self.upsertedCount = result.upsertedCount
+        if result.upsertedCount == 1 {
+            guard let id = result.upsertedIds[0] else {
+                throw RuntimeError.internalError(message: "BulkWriteResult missing _id for upserted document")
+            }
+            self.upsertedId = id
+        } else {
+            self.upsertedId = nil
+        }
+    }
+
+    private enum CodingKeys: String, CodingKey {
+        case matchedCount, modifiedCount, upsertedId, upsertedCount
+    }
+
+    public init(from decoder: Decoder) throws {
```

Comment:
I tried to just add the decodable conformance in the tests, but if you add an initializer in another module, it must delegate to an initializer from the defining module. so I would've had to add some kind of initializer here anyway that took in and set all of the properties. seemed not worth it, but open to doing so if anyone feels strongly.